pycqed/measurement/waveform_control/segment.py

```diff
@@ -810,15 +829,96 @@ def _test_overlap(self, allow_overlap=False, tol=1e-12):
+                            raise ValueError(msg)
+
+    def resolve_overlap(self):
+        """
+        Routine to resolve overlapping elements. Will be exectued if
+        corresponding pulsar parameter is true. This code
+        first goes through the list of overlapping elements that are
+        pairwise overlapping and clusters them into lists
+        of overlapping elements, where two different lists of overlapping
+        elements have no overlapping elements with
+        one another. At the end the code combines all elements of each
+        list into a new element.
+        """
+
+        self.gen_elements_on_awg()
+        self._test_overlap(track_and_ignore=True)
+        overlapping_elements = self.overlapping_elements
+
+        if len(overlapping_elements)==0:
+            return
+
+        # add first two overlapping elements to list
+        joint_overlapping_elements = [overlapping_elements[0]]
+
+        new_cluster = True
+        for i in range(len(overlapping_elements) - 1):
+            # check for all sets added to joint_overlapping_elements
+            # whether the next set of elements from
+            # overlapping_elements shares an element name. If so,
+            # then add this element to the entry in
+            # joint_overlapping_elements
+            for j in range(len(joint_overlapping_elements)):
+                if len(joint_overlapping_elements[j] & \
+                       overlapping_elements[i + 1]) != 0:
+                    joint_overlapping_elements[j] = \
+                        joint_overlapping_elements[j] | \
+                        overlapping_elements[i + 1]
+                    new_cluster = False
+
+            # if the new element from overlapping_elements overlaps
+            # with none of the previously added elements in
+            # joint_overlapping_elements (i.e. if new_cluster=True)
+            # add it as a new cluster.
+            if new_cluster:
+                joint_overlapping_elements.append(overlapping_elements[i + 1])
+            new_cluster = True
+
+        for i in range(len(joint_overlapping_elements)):
+            self._combine_elements(joint_overlapping_elements[i],
+                                   'overlapping_el_{}_{}'.format(i, self.name))
+
+
+    def _combine_elements(self, elements, combined_el_name):
+        """
+        Routine to properly combine elements in the segment.
+        :param elements: list or set of elements in the segment to be combined
+        :param combined_el_name: name of the combined element
+        :return:
+        """
+
+        new_pulse_list = []
+
+        for el in elements:
+            new_pulse_list+= self.elements.pop(el)
+            #remove it from element_start_end
+            self.element_start_end.pop(el)
+
+        # add new element
+        self.elements[combined_el_name] = new_pulse_list
+        # update new elements_on_awg
+        self.gen_elements_on_awg()
+        # update element_start_end
+
```

Comment:
please keep explaining comments together with the corresponding code, i.e., the blank line should be above this comment, not below.

Reply:
Done.